compiler/llvm/clangUtil.cpp

```diff
@@ -2966,6 +2966,70 @@ getCGArgInfo(const clang::CodeGen::CGFunctionInfo* CGI, int curCArg)
   return argInfo;
 }
 
+static unsigned helpGetCTypeAlignment(const clang::QualType& qType) {
+  GenInfo* info = gGenInfo;
+  INT_ASSERT(info);
+  ClangInfo* clangInfo = info->clangInfo;
+  INT_ASSERT(clangInfo);
+
+  unsigned alignInBits = clangInfo->Ctx->getTypeAlignIfKnown(qType);
+
+  unsigned alignInBytes = alignInBits / 8;
+  // round it up to a power of 2
+  unsigned rounded = 1;
+  while (rounded < alignInBytes) rounded *= 2;
+
+  return rounded;
+}
+
+static unsigned helpGetCTypeAlignment(const clang::TypeDecl* td) {
+  QualType qType;
+
+  if (const TypedefNameDecl* tnd = dyn_cast<TypedefNameDecl>(td)) {
+    qType = tnd->getCanonicalDecl()->getUnderlyingType();
+  } else if (const EnumDecl* ed = dyn_cast<EnumDecl>(td)) {
+    qType = ed->getCanonicalDecl()->getIntegerType();
+    // could also use getPromotionType()
+    //could also do:
+    //  qType =
+    //   tnd->getCanonicalDecl()->getTypeForDecl()->getCanonicalTypeInternal();
+  } else if (const RecordDecl* rd = dyn_cast<RecordDecl>(td)) {
+    RecordDecl *def = rd->getDefinition();
+    INT_ASSERT(def);
+    qType=def->getCanonicalDecl()->getTypeForDecl()->getCanonicalTypeInternal();
+  } else {
+    INT_FATAL("Unknown clang type declaration");
+  }
+
+  return helpGetCTypeAlignment(qType);
+}
+static unsigned helpGetAlignment(::Type* type) {
+  GenInfo* info = gGenInfo;
+  INT_ASSERT(info);
+
+  if (type->symbol->hasFlag(FLAG_EXTERN)) {
+    clang::TypeDecl* cType = NULL;
+    clang::ValueDecl* cVal = NULL;
+    info->lvt->getCDecl(type->symbol->cname, &cType, &cVal);
+    if (cType) {
+      return helpGetCTypeAlignment(cType);
+    }
+  }
+
+  // use the maximum alignment of all the fields
```

Comment:
This part gave me a pause. I don't think it is necessarily wrong, but I can't help but think that this is introducing an unnecessary alignment maybe? In other words, I am not sure that alignment of the fields in an aggregate type can be used to deduce the alignment of the type itself.

That being said, I can't imagine this being harmful in a noticeable way. (Maybe increased memory usage if you have arrays of aggregate types?)

Reply:
I would agree that it isn't necessarily optimal. Right now, alignment is only specifiable for `extern` types and only used for creating stack variables. I could instead make it an error if a field other than the first requires an alignment, for now. Does that sound good?

Reply:
I don't have a strong preference, as that may also be limiting to the user. I am fine with whatever call you make here.

Reply:
Hmm, Hello World doesn't compile with the error, so I will make a Future Work issue about it and add a couple of tests demonstrating.

Reply:
My understanding is that structure padding will ensure, if the structure starts at address 0, that everything is aligned appropriately. As a result, all that is necessary is to make sure that the structure is allocated to start out aligned to the maximum alignment necessary. I think the code here is correct but to see that we have to fix #17141 because the LLVM backend isn't emitting the structure padding at all right now.